app/src/main/java/org/greenstand/android/TreeTracker/models/messages/MessagesRepo.kt

```diff
@@ -146,24 +152,46 @@ class MessagesRepo(
             limit = query.limit,
         )
         query = result.query
+        if (query.total == 0) return@withContext
+
+        val saveMessageResponseJobs = mutableListOf<Job>()
+
+        result.messages.forEach {
+            ensureActive()
+            // launch all jobs in parallel and add these to a list.
+            saveMessageResponseJobs.add(launch {
+                saveMessageResponse(wallet, it.copy())
+            })
+        }
 
-        if (query.total == 0) return
+        // wait for all the jobs to complete
+        saveMessageResponseJobs.joinAll()
 
-        result.messages.forEach { saveMessageResponse(wallet, it) }
         while (query.total >= query.limit + query.offset) {
+            ensureActive()
             query = result.query.copy(offset = result.query.offset + result.query.limit)
             result = apiService.getMessages(
                 wallet = wallet,
                 lastSyncTime = lastSyncTime,
                 offset = query.offset,
                 limit = query.limit,
             )
-            result.messages.forEach { saveMessageResponse(wallet, it) }
+
+            saveMessageResponseJobs.clear()
+
+            result.messages.forEach {
+                ensureActive()
+                saveMessageResponseJobs.add(launch {
+                    saveMessageResponse(wallet, it)
+                })
```

Comment:
Would using async here make more sense?
saveMessageResponseJobs += async { saveMessageResponse(wallet, it) }

Reply:
https://stackoverflow.com/questions/46226518/what-is-the-difference-between-launch-join-and-async-await-in-kotlin-coroutines here's a sof question, the saveMessageResponse is a fire and forget task hence launch is used here.